### DISCUSSION TITLE
Uninstall Applications automatically when uninstalling Argo

### DISCUSSION THREAD

**Discussion by mgs-garcia** [argoproj/argo-helm#1744] [Q&A] [answered]
Hi, is there any way to configure in Argo that if it is uninstalled, then all the Applications that are handled by Argo also get uninstalled?
I thought about this because of an issue I'm facing... I have configured Argo in my Terraform IaC configuration so it is automatically deployed on my kubernetes cluster when it is provisioned. Then I install some Applications (manually) with Argo UI, which in turn spin up multiple kubernetes resources. Imagine that after some time I want to remove the kubernetes cluster, therefore I invoke `terraform destroy` which will first remove Argo from the kubernetes cluster. The problem is that Argo cannot be completely removed because it has Application resources inside. Plus, those Applications also created related infrastructure (i.e. load balancers) that should be removed as the kubernetes cluster is meant to be destroyed, otherwise they would be left hanging.

I don't think there's an option in the Argo Helm Chart that when it detects the uninstall command, all `Application` resources are removed first. Would it make sense to include that in a future release? Thanks

**Comment by pdrastil** [comment, marked as answer]
Hi, I believe this is not possible as uninstalling Argo CD will remove controllers responsible for finalisation of managed resources and for this reason it's not recommended to deploy / manage your applications resources together with Argo CD. I would instead use another chart with app of apps pattern (i.e. argocd-apps or one of your making) that you will uninstall first and wait until Argo CD do the cleanup. Then you can throw away / uninstall Argo CD safely.

**Reply by MonicaMagoniCom** [reply]
How did you achieve it? 

```
resource "helm_release" "argocd" {
  name  = "argocd"

  repository       = "https://argoproj.github.io/argo-helm"
  chart            = "argo-cd"
  namespace        = "argocd"
  version          = "6.7.2"
  create_namespace = true
  values = [
    file("argocd/values.yaml")
  ]
}

resource "helm_release" "argocd-apps" {
  name  = "argocd-apps"

  repository       = "https://argoproj.github.io/argo-helm"
  chart            = "argocd-apps"
  namespace        = "argocd"
  version          = "1.6.2"

  values = [
    file("argocd/application.yaml")
  ]

  depends_on = [helm_release.argocd]
}

```

where 'argocd/application.yaml' implemented the app of apps pattern. But when I do terraform destroy, application are not destroyed.

**Reply by pdrastil** [reply]
@MonicaMagoniCom - This looks correct. Do you set finalizer in managed apps? By default Application resource is not removing from cluster without this.

Ref: https://argo-cd.readthedocs.io/en/stable/user-guide/app_deletion/#about-the-deletion-finalizer
